lab_2/main.py

```diff
@@ -6,14 +6,31 @@
 from lab_1.main import tokenize, remove_stop_words
 
 
+def elements_instances(iterable, *types):
+    """
+    Checks if all elements in an iterable are instances of allowed types
+    :param iterable: an iterable with elements to be checked
+    :params types: types allowed to be in the iterable
+    """
+    return all(any(isinstance(elem, t) for t in types) for elem in iterable)
+
 # 4
 def get_freq_dict(tokens: list) -> dict or None:
     """
     Calculates frequencies of given tokens
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+    if not isinstance(tokens, list):
+        return None
+    if not elements_instances(tokens, str):
+        return None
+    freq_dict = {}
+    for token in tokens:
+        if token not in freq_dict:
+            freq_dict[token] = 0
+        freq_dict[token] += 1 / len(tokens)
+    return {k: round(v, 5) for k, v in freq_dict.items()}
```

Comment:
why to generate a dictionary once more? maybe you can round earlier?

Reply:
In order to avoid increasing the error bars on the result, rounding must happen once, rather than after every addition. This necessitates a follow-up iteration over the dictionary, following the counting iteration.